### PR TITLE
feat: query product releases (Cr 24615)

### PR DESCRIPTION
**Pull request by kim-codefresh**

### CODE REVIEW

File: docs/commands/cli-v2_product-release_list.md [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+```
+
+            cli-v2 product-release list --product <product>
+        
+```
+
+### Options
+
+```
+  -e, --end-date string               Filter by end date
```
Comment (ATGardner) [inline docs/commands/cli-v2_product-release_list.md]: how does this filter works? everything `>`  than enddate? `>=`? `<`? `<=`? maybe just be clear in the description

File: docs/commands/cli-v2_product-release_list.md [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+  -h, --help                          help for list
+  -p, --product string                product
+      --promotion-flows stringArray   Filter by promotion flows
+  -s, --start-date string             Filter by start date
```
Comment (ATGardner) [inline docs/commands/cli-v2_product-release_list.md]: same

File: docs/commands/cli-v2_product-release_list.md [hunk 3]
```diff
@@ -0,0 +1,41 @@
+## cli-v2 product-release list
+
+List all the pipelines
```
Comment (ATGardner) [inline docs/commands/cli-v2_product-release_list.md]: should be "product releases"

also - explicitly mention how many results are returned by default (when no start/end date are entered.

File: docs/commands/cli-v2_product-release_list.md (hunk 1 above)
Comment (kim-codefresh) [reply docs/commands/cli-v2_product-release_list.md]: oh this one actually no relevant in the end, forgot to delete it

File: cmd/commands/product-release.go [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+)
+
+type (
+	Slice struct {
+		Edges    []*Edge              `json:"edges"`
+		PageInfo *platmodel.SliceInfo `json:"pageInfo"`
+	}
+
+	Edge struct {
+		Node map[string]any `json:"node"`
```
Comment (ATGardner) [inline cmd/commands/product-release.go:36]: 👍🏻

File: docs/commands/cli-v2_product-release_list.md (hunk 1 above)
Comment (kim-codefresh) [reply docs/commands/cli-v2_product-release_list.md]: I guess I didnt regenerated the docs

File: cmd/commands/product-release.go [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+			}
+			return RunProductReleaseList(ctx, filterArgs, productName, page, pageLimit)
+		},
+	}
+
+	cmd.Flags().StringVarP(&statusList, "status", "s", "", "Filter by statuses")
+	cmd.Flags().StringVar(&promotionFlowList, "promotion-flows", "", "Filter by promotion flows")
+	cmd.Flags().IntVar(&page, "page", 1, "page number")
+	cmd.Flags().IntVar(&pageLimit, "page-limit", 20, "page limit number")
+	cmd.Flags().StringVarP(&productName, "product", "p", "", "product")
```
Comment (ATGardner) [inline cmd/commands/product-release.go]: where are the start and end date filters?

File: cmd/commands/product-release.go [hunk 6]
```diff
@@ -0,0 +1,245 @@
...
+		edges {
+		  node {
+			releaseId
+			steps {
+			  environmentName
+			  status
+			}
+			status
+		  }
+		}
```
Comment (ATGardner) [inline cmd/commands/product-release.go]: all those lines should be indented one more time to the right

File: cmd/commands/product-release.go (hunk 5 above)
Comment (kim-codefresh) [reply cmd/commands/product-release.go]: yeah we dont use them in the end I just forgot to generate the docs, sorry for confusion

Comment (ATGardner) [reply cmd/commands/product-release.go]: and why don't i see pageLimit in the md file?

File: cmd/commands/product-release.go [hunk 7]
```diff
@@ -0,0 +1,245 @@
...
+  `
+	// add pagination - default for now is last 20
+	variables := map[string]any{
+		"filters":     filterArgs,
+		"productName": productName,
+		"pagination":  pagination,
+	}
+
+	var productReleasesPage map[string]any
+	productReleasesPage, err := client.GraphqlAPI[map[string]any](ctx, cfConfig.NewClient().InternalClient(), query, variables)
```
Comment (ATGardner) [inline cmd/commands/product-release.go]: i think that will do
```suggestion
	productReleasesPage, err := client.GraphqlAPI[map[string]any](ctx, cfConfig.NewClient().InternalClient(), query, variables)
```

File: cmd/commands/product-release.go [hunk 8]
```diff
@@ -0,0 +1,245 @@
...
+	if err != nil {
+		return nil, err
+	}
+
+	return &Slice{
+		Edges:    edges,
+		PageInfo: pageInfo,
+	}, nil
+}
+func ToProductReleaseStatus(stringStatusList string) ([]platmodel.ProductReleaseStatus, error) {
```
Comment (ATGardner) [inline cmd/commands/product-release.go]: add an empty line between funcs

File: cmd/commands/product-release.go [hunk 9]
```diff
@@ -0,0 +1,245 @@
...
+		"productName": productName,
+		"pagination":  pagination,
+	}
+
+	var productReleasesPage map[string]any
+	productReleasesPage, err := client.GraphqlAPI[map[string]any](ctx, cfConfig.NewClient().InternalClient(), query, variables)
+	if err != nil {
+		return err
+	}
+	productReleaseSlice, err := TransformSlice(productReleasesPage)
```
Comment (ATGardner) [inline cmd/commands/product-release.go]: need to handle the err in here

File: cmd/commands/product-release.go [hunk 10]
```diff
@@ -0,0 +1,245 @@
...
+	if len(productReleaseSlice.Edges) == 0 {
+		return fmt.Errorf("no product releases found")
+	}
+	resJSON, err := json.MarshalIndent(productReleaseSlice.Edges, "", "\t")
```
Comment (ATGardner) [inline cmd/commands/product-release.go]: and here

File: cmd/commands/product-release.go (hunk 7 above)
Comment (ATGardner) [reply cmd/commands/product-release.go]: also - shouldn't this just return the `Slice`?

File: cmd/commands/product-release.go [hunk 11]
```diff
@@ -0,0 +1,245 @@
...
+	hasPrevPage, _ := data["hasPrevPage"].(bool)
+	return &platmodel.SliceInfo{
+		StartCursor: &startCursor,
+		EndCursor:   &endCursor,
+		HasNextPage: hasNextPage,
+		HasPrevPage: hasPrevPage,
+	}, nil
+}
+
+func TransformSlice(slice map[string]any) (*Slice, error) {
```
Comment (ATGardner) [inline cmd/commands/product-release.go]: yeah - i don't think you need that at all

Comment (ATGardner) [reply cmd/commands/product-release.go]: if the generic graphql call will return a `Slice`, it will do all this for you

Comment (kim-codefresh) [reply cmd/commands/product-release.go]: what would you do instead?

File: cmd/commands/product-release.go [hunk 12]
```diff
@@ -0,0 +1,245 @@
...
+	productReleasesPage, err := client.GraphqlAPI[map[string]any](ctx, cfConfig.NewClient().InternalClient(), query, variables)
+	if err != nil {
+		return err
+	}
+	productReleaseSlice, err := TransformSlice(productReleasesPage)
+	if len(productReleaseSlice.Edges) == 0 {
+		return fmt.Errorf("no product releases found")
+	}
+	resJSON, err := json.MarshalIndent(productReleaseSlice.Edges, "", "\t")
+
```
Comment (ATGardner) [inline cmd/commands/product-release.go:133]: wouldn't that also print out "node" for every release? shouldn't you iterate over the edges, and turn them into a slice of productReleases (which are `map[string]any`)

File: cmd/commands/product-release.go [hunk 13]
```diff
@@ -0,0 +1,245 @@
...
+			}
+			filterArgs := platmodel.ProductReleaseFiltersArgs{
+				Statuses:       releaseStatus,
+				PromotionFlows: promotionFlows,
+			}
+			return RunProductReleaseList(ctx, filterArgs, productName, page, pageLimit)
+		},
+	}
+
+	cmd.Flags().StringVarP(&statusList, "status", "s", "", "Filter by statuses")
```
Comment (ATGardner) [inline cmd/commands/product-release.go]: i think there is a flag type for []string as well

File: cmd/commands/product-release.go [hunk 14]
```diff
@@ -0,0 +1,245 @@
...
+		productName       string
+		page              int
+		pageLimit         int
+		statusList        string
+		promotionFlowList string
+	)
+
+	cmd := &cobra.Command{
+		Use:   "list",
+		Short: "List all the pipelines",
```
Comment (eti-codefresh) [inline cmd/commands/product-release.go]: all the product releases

File: cmd/commands/product-release.go [hunk 15]
```diff
@@ -0,0 +1,245 @@
...
+		Args:  cobra.NoArgs,
+		Example: util.Doc(`
+			<BIN> product-release list --product <product>
```
Comment (eti-codefresh) [inline cmd/commands/product-release.go]: is there indication that we have some limit?

File: docs/commands/cli-v2_product-release.md [hunk 16]
```diff
@@ -0,0 +1,29 @@
...
+      --cfconfig string            Custom path for authentication contexts config file (default "/home/user")
+      --insecure                   Disable certificate validation for TLS connections (e.g. to g.codefresh.io)
+      --insecure-ingress-host      Disable certificate validation of ingress host (default: false)
+      --request-timeout duration   Request timeout (default 30s)
+```
+
+### SEE ALSO
+
+* [cli-v2](cli-v2.md)	 - cli-v2 is used for installing and managing codefresh installations using gitops
+* [cli-v2 product-release list](cli-v2_product-release_list.md)	 - List all the pipelines
```
Comment (eti-codefresh) [inline docs/commands/cli-v2_product-release.md]: maybe add here that the result is limited

File: cmd/commands/product-release.go (hunk 15 above)
Comment (ATGardner) [reply cmd/commands/product-release.go]: is product a mandatory field?
if it is - switch it into a positional argument, and mark it as mandatory in cobra
if it isn't - make it clear in the example (by putting it in `[--product product]`

File: cmd/commands/product-release.go [hunk 17]
```diff
@@ -0,0 +1,173 @@
...
+		`),
+		RunE: func(cmd *cobra.Command, args []string) error {
+			ctx := cmd.Context()
+			releaseStatus, err := toProductReleaseStatus(statuses)
+			if err != nil {
+				return fmt.Errorf("failed to convert status: %w", err)
+			}
+			if len(args) == 0 {
+				return fmt.Errorf("missing product name")
+			}
```
Comment (ATGardner) [inline cmd/commands/product-release.go]: put that check first. it is "more" required

File: cmd/commands/product-release.go [hunk 18]
```diff
@@ -0,0 +1,173 @@
...
+	)
+
+	cmd := &cobra.Command{
+		Use:   "list",
+		Short: "List all product releases",
+		Args:  cobra.MaximumNArgs(1),
+		Example: util.Doc(`
+			<BIN> product-release list <product-name>
+			<BIN> product-release list product-name --page-limit 3
+			<BIN> product-release list product-name --status RUNNING,FAILED --promotion-flows base-flow,flow-2
```
Comment (ATGardner) [inline cmd/commands/product-release.go]:
```suggestion
			<BIN> product-release list <product-name> --page-limit 3
			<BIN> product-release list <product-name> --status RUNNING,FAILED --promotion-flows base-flow,flow-2
```

**Review by ATGardner — APPROVED** [review]

**Pull request merged**